File: tree/easy/bfs_dfs/merge_two_binary_trees_617.go
package bfs_dfs

import "github.com/vielendanke/algorithms_in_place/tree"

func mergeTrees(root1 *tree.TreeNode, root2 *tree.TreeNode) *tree.TreeNode {
	// handle null nodes
	// handle sum of nodes
	// start from the root -> perform root merge -> do the same for left / right node
	if root1 == nil {
		return root2
	}
	if root2 == nil {
		return root1
	}
	root1.Val += root2.Val
	root1.Left = mergeTrees(root1.Left, root2.Left)
	root1.Right = mergeTrees(root1.Right, root2.Right)
	return root1
}

func mergeTreesIterative(root1 *tree.TreeNode, root2 *tree.TreeNode) *tree.TreeNode {
	// handle null nodes
	// handle sum of nodes
	// start from the root -> perform root merge -> do the same for left / right node
	if root1 == nil {
		return root2
	}
	stack := make([][]*tree.TreeNode, 0)

	stack = append(stack, []*tree.TreeNode{root1, root2})

	var tempRootPair []*tree.TreeNode

	for len(stack) > 0 {
		tempRootPair, stack = stack[len(stack)-1], stack[:len(stack)-1]
		if tempRootPair[0] == nil || tempRootPair[1] == nil {
			continue
		}
		tempRootPair[0].Val += tempRootPair[1].Val

		if tempRootPair[0].Left == nil {
			tempRootPair[0].Left = tempRootPair[1].Left
		} else {
			stack = append(stack, []*tree.TreeNode{tempRootPair[0].Left, tempRootPair[1].Left})
		}
		if tempRootPair[0].Right == nil {
			tempRootPair[0].Right = tempRootPair[1].Right
		} else {
			stack = append(stack, []*tree.TreeNode{tempRootPair[0].Right, tempRootPair[1].Right})
		}
	}
	return root1
}

File: tree/easy/java_solutions/MergeTwoBinaryTrees_617.java
package tree.easy.java_solutions;

import tree.Helper;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class MergeTwoBinaryTrees_617 {

    public Helper.TreeNode mergeTrees(Helper.TreeNode root1, Helper.TreeNode root2) {
        if (root1 == null) {
            return root2;
        }
        if (root2 == null) {
            return root1;
        }
        root1.val += root2.val;
        root1.left = mergeTrees(root1.left, root2.left);
        root1.right = mergeTrees(root1.right, root2.right);
        return root1;
    }

    public Helper.TreeNode mergeTreesUsingStack(Helper.TreeNode root1, Helper.TreeNode root2) {
        if (root1 == null) {
            return root2;
        }
        Stack<List<Helper.TreeNode>> stack = new Stack<>();

        stack.add(new ArrayList<>() {{
            add(root1);
            add(root2);
        }});

        while (stack.size() > 0) {
            List<Helper.TreeNode> nodes = stack.pop();
            if (nodes.get(0) == null || nodes.get(1) == null) {
                continue;
            }
            nodes.get(0).val += nodes.get(1).val;

            if (nodes.get(0).left == null) {
                nodes.get(0).left = nodes.get(1).left;
            } else {
                stack.add(new ArrayList<>() {{
                    add(nodes.get(0).left);
                    add(nodes.get(1).left);
                }});
            }
            if (nodes.get(0).right == null) {
                nodes.get(0).right = nodes.get(1).right;
            } else {
                stack.add(new ArrayList<>() {{
                    add(nodes.get(0).right);
                    add(nodes.get(1).right);
                }});
            }
        }
        return root1;
    }
}
